### ISSUE TITLE
Can't write; duplicate key in table

### ISSUE THREAD

**Issue by alan345** [prisma/upgrade#46]
Recommandation from prisma-upgrade:

```
    ALTER TABLE `Contract` ADD COLUMN `signedById` char(25) CHARACTER SET utf8;
    UPDATE `Contract`, `_UserContractSignedBy` SET `Contract`.`signedById` = `_UserContractSignedBy`.B where `_UserContractSignedBy`.A = `Contract`.`id`;
    ALTER TABLE `Contract` ADD CONSTRAINT signedBy FOREIGN KEY (`signedById`) REFERENCES `User`(`id`);
    DROP TABLE `_UserContractSignedBy`;
    ALTER TABLE `Contract` ADD COLUMN `signedByVendorId` char(25) CHARACTER SET utf8;
    UPDATE `Contract`, `_UserContractSignedByVendor` SET `Contract`.`signedByVendorId` = `_UserContractSignedByVendor`.B where `_UserContractSignedByVendor`.A = `Contract`.`id`;
    ALTER TABLE `Contract` ADD CONSTRAINT signedByVendor FOREIGN KEY (`signedByVendorId`) REFERENCES `User`(`id`);
    DROP TABLE `_UserContractSignedByVendor`;
    ALTER TABLE `Contract` ADD COLUMN `createdById` char(25) CHARACTER SET utf8;
    UPDATE `Contract`, `_ContractCreatedBy` SET `Contract`.`createdById` = `_ContractCreatedBy`.B where `_ContractCreatedBy`.A = `Contract`.`id`;
    ALTER TABLE `Contract` ADD CONSTRAINT createdBy FOREIGN KEY (`createdById`) REFERENCES `User`(`id`);
    DROP TABLE `_ContractCreatedBy`;
    ALTER TABLE `Cardholder` ADD COLUMN `userId` char(25) CHARACTER SET utf8;
    UPDATE `Cardholder`, `_CardholderUserRelation` SET `Cardholder`.`userId` = `_CardholderUserRelation`.B where `_CardholderUserRelation`.A = `Cardholder`.`id`;
    ALTER TABLE `Cardholder` ADD CONSTRAINT user FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
    DROP TABLE `_CardholderUserRelation`;
    ALTER TABLE `NotificationUserIssuedCard` ADD COLUMN `userId` char(25) CHARACTER SET utf8;
    UPDATE `NotificationUserIssuedCard`, `_NotificationUserIssuedCardRelation` SET `NotificationUserIssuedCard`.`userId` = `_NotificationUserIssuedCardRelation`.B where `_NotificationUserIssuedCardRelation`.A = `NotificationUserIssuedCard`.`id`;
    ALTER TABLE `NotificationUserIssuedCard` CHANGE `userId` `userId` char(25) CHARACTER SET utf8 NOT NULL;
    ALTER TABLE `NotificationUserIssuedCard` ADD CONSTRAINT user FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
    DROP TABLE `_NotificationUserIssuedCardRelation`;

```

The result:
```
mysql>   ALTER TABLE `Contract` ADD COLUMN `signedById` char(25) CHARACTER SET utf8;
Query OK, 0 rows affected (0.38 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql>     UPDATE `Contract`, `_UserContractSignedBy` SET `Contract`.`signedById` = `_UserContractSignedBy`.B where `_UserContractSignedBy`.A = `Contract`.`id`;
Query OK, 10 rows affected (0.01 sec)
Rows matched: 10  Changed: 10  Warnings: 0

mysql>     ALTER TABLE `Contract` ADD CONSTRAINT signedBy FOREIGN KEY (`signedById`) REFERENCES `User`(`id`);
Query OK, 91 rows affected (0.14 sec)
Records: 91  Duplicates: 0  Warnings: 0

mysql>     DROP TABLE `_UserContractSignedBy`;
Query OK, 0 rows affected (0.01 sec)

mysql>     ALTER TABLE `Contract` ADD COLUMN `signedByVendorId` char(25) CHARACTER SET utf8;
Query OK, 0 rows affected (0.27 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql>     UPDATE `Contract`, `_UserContractSignedByVendor` SET `Contract`.`signedByVendorId` = `_UserContractSignedByVendor`.B where `_UserContractSignedByVendor`.A = `Contract`.`id`;
Query OK, 12 rows affected (0.01 sec)
Rows matched: 12  Changed: 12  Warnings: 0

mysql>     ALTER TABLE `Contract` ADD CONSTRAINT signedByVendor FOREIGN KEY (`signedByVendorId`) REFERENCES `User`(`id`);
Query OK, 91 rows affected (0.15 sec)
Records: 91  Duplicates: 0  Warnings: 0

mysql>     DROP TABLE `_UserContractSignedByVendor`;
Query OK, 0 rows affected (0.01 sec)

mysql>   ALTER TABLE `Contract` ADD COLUMN `createdById` char(25) CHARACTER SET utf8;
Query OK, 0 rows affected (0.24 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql>     UPDATE `Contract`, `_ContractCreatedBy` SET `Contract`.`createdById` = `_ContractCreatedBy`.B where `_ContractCreatedBy`.A = `Contract`.`id`;
Query OK, 91 rows affected (0.02 sec)
Rows matched: 91  Changed: 91  Warnings: 0

mysql>     ALTER TABLE `Contract` ADD CONSTRAINT createdBy FOREIGN KEY (`createdById`) REFERENCES `User`(`id`);
Query OK, 91 rows affected (0.15 sec)
Records: 91  Duplicates: 0  Warnings: 0

mysql>     DROP TABLE `_ContractCreatedBy`;
Query OK, 0 rows affected (0.02 sec)

mysql>     ALTER TABLE `Cardholder` ADD COLUMN `userId` char(25) CHARACTER SET utf8;
Query OK, 0 rows affected (0.06 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql>     UPDATE `Cardholder`, `_CardholderUserRelation` SET `Cardholder`.`userId` = `_CardholderUserRelation`.B where `_CardholderUserRelation`.A = `Cardholder`.`id`;
Query OK, 174 rows affected (0.03 sec)
Rows matched: 174  Changed: 174  Warnings: 0

mysql>     ALTER TABLE `Cardholder` ADD CONSTRAINT user FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
Query OK, 174 rows affected (0.05 sec)
Records: 174  Duplicates: 0  Warnings: 0

mysql>     DROP TABLE `_CardholderUserRelation`;
Query OK, 0 rows affected (0.02 sec)

mysql>   ALTER TABLE `NotificationUserIssuedCard` ADD COLUMN `userId` char(25) CHARACTER SET utf8;
Query OK, 0 rows affected (0.07 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql>     UPDATE `NotificationUserIssuedCard`, `_NotificationUserIssuedCardRelation` SET `NotificationUserIssuedCard`.`userId` = `_NotificationUserIssuedCardRelation`.B where `_NotificationUserIssuedCardRelation`.A = `NotificationUserIssuedCard`.`id`;
Query OK, 12 rows affected (0.01 sec)
Rows matched: 12  Changed: 12  Warnings: 0

mysql>     ALTER TABLE `NotificationUserIssuedCard` CHANGE `userId` `userId` char(25) CHARACTER SET utf8 NOT NULL;
Query OK, 0 rows affected (0.06 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql>     ALTER TABLE `NotificationUserIssuedCard` ADD CONSTRAINT user FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
ERROR 1022 (23000): Can't write; duplicate key in table '#sql-1_a25'
```

Sometime we get the error: `ERROR 1022 (23000): Can't write; duplicate key in table '#sql-1_a25'`

Should we skip this error? or something should be fixed?

**Comment by matthewmueller** [comment]
Hmm, interesting. 

- What is `#sql-1_9f9`? A table? Some value in the `userId` column? 
- Are there duplicate keys in `userId`?

**Comment by alan345** [comment]
- No idea.. Nope, not table, not a value either.. 
- nope. It is a prisma1 model, It doesn't have duplicate.



This is the table `NotificationUserIssuedCard` before running the SQLs queries:

```
mysql> Describe NotificationUserIssuedCard;
+--------------------+-------------+------+-----+---------+-------+
| Field              | Type        | Null | Key | Default | Extra |
+--------------------+-------------+------+-----+---------+-------+
| id                 | char(25)    | NO   | PRI | NULL    |       |
| createdAt          | datetime(3) | NO   |     | NULL    |       |
| emailNotifications | tinyint(1)  | NO   |     | NULL    |       |
| updatedAt          | datetime(3) | NO   |     | NULL    |       |
+--------------------+-------------+------+-----+---------+-------+
4 rows in set (0.01 sec)
```
and after
```
mysql> describe NotificationUserIssuedCard;
+--------------------+-------------+------+-----+---------+-------+
| Field              | Type        | Null | Key | Default | Extra |
+--------------------+-------------+------+-----+---------+-------+
| id                 | char(25)    | NO   | PRI | NULL    |       |
| createdAt          | datetime(3) | NO   |     | NULL    |       |
| emailNotifications | tinyint(1)  | NO   |     | NULL    |       |
| updatedAt          | datetime(3) | NO   |     | NULL    |       |
| userId             | char(25)    | NO   |     | NULL    |       |
+--------------------+-------------+------+-----+---------+-------+
5 rows in set (0.01 sec)
```

Also, here they are talking about `Constraints must be unique for the entire database, not just for the specific table you are creating/altering.` https://stackoverflow.com/questions/18056786/error-1022-cant-write-duplicate-key-in-table#:~:text=The%20most%20likely%20you%20already,table%20you%20are%20creating%2Faltering.&text=Change%20the%20Foreign%20key%20name%20in%20MySQL.

**Comment by alan345** [comment]
I hope it is not because all our external keys should have a unique name?
example: 

```
type NotificationUserIssuedCard {
  id: ID! @unique
  emailNotifications: Boolean! 
  user: User! @relation(name: "NotificationUserIssuedCardRelation") # HERE THE NAME should be unique?
}

type Cardholder {
  id: ID! @unique
  name: String
  user: User! @relation(name: "CardholderUserRelation") # HERE THE NAME should be unique?
}

type User {
  id: ID! @unique
  name: String
}
```

**Comment by matthewmueller** [comment]
Thanks for the helpful information!

I think it's related to the constraint name not being unique, could you try...

```sql
ALTER TABLE `NotificationUserIssuedCard` ADD CONSTRAINT NotificationUserIssuedCard_userId_fk FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
```

?

**Comment by alan345** [comment]
Yep it works.
```
mysql>     ALTER TABLE `NotificationUserIssuedCard` ADD CONSTRAINT user FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
ERROR 1022 (23000): Can't write; duplicate key in table '#sql-1_f6'
mysql> ALTER TABLE `NotificationUserIssuedCard` ADD CONSTRAINT NotificationUserIssuedCard_userId_fk FOREIGN KEY (`userId`) REFERENCES `User`(`id`);
Query OK, 12 rows affected (0.05 sec)
Records: 12  Duplicates: 0  Warnings: 0
```

**Comment by matthewmueller** [comment]
Great, thanks for testing it! I'll fix this tomorrow.

**Comment by matthewmueller** [comment]
Learned my lesson again to never commit to a date 😅 

Should be fixed in `0.0.30`.

Sorry for the delay @alan345!
